### PR TITLE
[MQTT] Wait until removing the retained message. @open sesame 08/23 13:10

### PR DESCRIPTION
**Pull request by gichan-jang**
Change to wait until removing the retained message.

### CODE REVIEW

File: src/libnnstreamer-edge/nnstreamer-edge-mqtt-mosquitto.c [hunk 1]
```diff
@@ -38,6 +38,10 @@ typedef struct
   /* event callback for new message */
   nns_edge_event_cb event_cb;
   void *user_data;
+
+  pthread_mutex_t lock;
+  pthread_cond_t cond;
```
Comment (songgot) [inline src/libnnstreamer-edge/nnstreamer-edge-mqtt-mosquitto.c:43]: is it necessary?

Comment (songgot) [reply src/libnnstreamer-edge/nnstreamer-edge-mqtt-mosquitto.c:43]: Sorry, I got it

File: src/libnnstreamer-edge/nnstreamer-edge-mqtt-mosquitto.c [hunk 2]
```diff
@@ -214,6 +218,26 @@ nns_edge_mqtt_connect (const char *id, const char *topic, const char *host,
   return ret;
 }
 
+/**
+ * @brief publish callback.
+ * @note This callback called both if the message is sent successfully or if the broker responded with an error.
```
Comment (songgot) [inline src/libnnstreamer-edge/nnstreamer-edge-mqtt-mosquitto.c:223]: Typo (called -> is called) ?

**Review by songgot — APPROVED** [review]
LGTM, Thank you!

File: src/libnnstreamer-edge/nnstreamer-edge-mqtt-mosquitto.c (hunk 2 above)
Comment (gichan-jang) [reply src/libnnstreamer-edge/nnstreamer-edge-mqtt-mosquitto.c]: Updated.

**Review by jaeyun-jung — APPROVED** [review]

File: src/libnnstreamer-edge/nnstreamer-edge-mqtt-mosquitto.c [hunk 3]
```diff
@@ -214,6 +221,26 @@ nns_edge_mqtt_connect (const char *id, const char *topic, const char *host,
...
+  nns_edge_broker_s *bh = NULL;
+
+  bh = (nns_edge_broker_s *) mosquitto_userdata (mosq);
+
+  if (!bh)
+    return;
+
+  nns_edge_lock (bh);
+  bh->cleared = true;
+  nns_edge_cond_signal (bh);
```
Comment (myungjoo) [inline src/libnnstreamer-edge/nnstreamer-edge-mqtt-mosquitto.c:240]: add ```if (bh->cond)``` (refer to the next comment)

File: src/libnnstreamer-edge/nnstreamer-edge-mqtt-mosquitto.c [hunk 4]
```diff
@@ -236,9 +263,18 @@ nns_edge_mqtt_close (nns_edge_broker_h broker_h)
...
+    mosquitto_publish_callback_set (handle, _publish_cb);
     mosquitto_publish (handle, NULL, bh->topic, 0, NULL, 1, true);
 
+    nns_edge_lock (bh);
+    if (!bh->cleared) {
+      nns_edge_cond_wait_until (bh, 1000U);
+    }
+    nns_edge_unlock (bh);
+    nns_edge_lock_destroy (bh);
+    nns_edge_cond_destroy (bh);
```
[the remark this answers is not in the document]
Comment (gichan-jang) [reply src/libnnstreamer-edge/nnstreamer-edge-mqtt-mosquitto.c]: Thanks. Updated.

**Pull request merged**